File: 004-queue/queue.js
class Queue {
  constructor(capacity) {
    this.list = new Array(capacity);
    this.head = -1;
    this.tail = -1;
  }

  isQueueFull() {
    return this.head === (this.tail + 1) % this.list.length;
  }

  isQueueEmpty() {
    return this.head === -1;
  }

  enqueue(val) {
    if (this.isQueueFull()) {
      throw new Error("overflow");
    }
    if (this.head === -1) {
      this.head = 0;
    }
    this.tail = (this.tail + 1) % this.list.length;
    this.list[this.tail] = val;
    return val;
  }

  dequeue() {
    if (this.isQueueEmpty()) {
      throw new Error("undeflow");
    }
    const x = this.list[this.head];
    if (this.head === this.tail) {
      this.head = this.tail = -1;
    } else {
      this.head = (this.head + 1) % this.list.length;
    }
    return x;
  }

  traverse() {
    let start = this.head;
    while (start !== this.tail) {
      console.log(this.list[start]);
      start = (start + 1) % this.list.length;
    }
    console.log(this.list[start]);
  }
}

const queue = new Queue(5);

try {
  console.log("dequeue " + queue.dequeue());
} catch (e) {
  console.log("dequeue error ", e);
}
console.log("enqueue " + queue.enqueue(12));
try {
  console.log("dequeue " + queue.dequeue());
} catch (e) {
  console.log("dequeue error ", e);
}
try {
  console.log("dequeue " + queue.dequeue());
} catch (e) {
  console.log("dequeue error ", e);
}

[2, 4, 3, 6, 7, 5, 6].forEach(val => {
  try {
    console.log("enqueue " + queue.enqueue(val));
  } catch (e) {
    console.log("enqueue error : ", val, e);
  }
});

console.log("queue traverse() ");
queue.traverse();

for (let i = 0; i < 8; i++) {
  try {
    console.log("dequeue " + queue.dequeue());
  } catch (e) {
    console.log("dequeue error ", e);
  }
}

File: 004-queue/queue.go
package main

import (
	"errors"
	"fmt"
)

type queue struct {
	list []int
	head int
	tail int
}

func (q *queue) isQueueEmpty() bool {
	return q.head == -1
}

func (q *queue) isQueueFull() bool {
	return q.head == (q.tail+1)%len(q.list)
}

func (q *queue) enqueue(val int) (int, error) {
	if q.isQueueFull() {
		return -1, errors.New("overflow")
	}
	if q.head == -1 {
		q.head = 0
	}
	q.tail = (q.tail + 1) % len(q.list)
	q.list[q.tail] = val
	return val, nil
}

func (q *queue) dequeue() (int, error) {
	if q.isQueueEmpty() {
		return -1, errors.New("underflow")
	}
	x := q.list[q.head]
	if q.head == q.tail {
		q.head = -1
		q.tail = -1
	} else {
		q.head = (q.head + 1) % len(q.list)
	}
	return x, nil
}

func (q *queue) traverse() []int {
	start := q.head
	result := []int{}
	for start != q.tail {
		result = append(result, q.list[start])
		start = (start + 1) % len(q.list)
	}
	result = append(result, q.list[start])
	return result
}

func new(capacity int) queue {
	q := queue{
		list: make([]int, capacity, capacity),
		head: -1,
		tail: -1,
	}
	return q
}

func main() {
	q := new(5)

	q.enqueue(5)
	q.enqueue(8)
	q.enqueue(6)
	q.enqueue(2)
	q.enqueue(9)
	fmt.Println("queue", q.traverse())
	_, err := q.enqueue(11)
	if err != nil {
		fmt.Println(err)
	}
	for i := 1; i < 8; i++ {
		v, err := q.dequeue()
		if err != nil {
			fmt.Println(err)
		} else {
			fmt.Println("dequeue ", v)
		}
	}
	q.enqueue(23)
	q.enqueue(12)
	q.enqueue(23)
	fmt.Println("queue", q.traverse())
	v, err := q.dequeue()
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("dequeue ", v)
	}
	fmt.Println("queue", q.traverse())
}
